_template_fastapi_/provider/app.py

```diff
@@ -0,0 +1,58 @@
+import logging
+import os
+import uvicorn
+
+from client import CustomClient
+from typing import List, Optional
+from dotenv import load_dotenv
+from fastapi import Depends, FastAPI, HTTPException, Header, Request
+from datamodels import DataItem, Request, Response
+
+
+logging.basicConfig(level=logging.DEBUG)
+
+load_dotenv()
+SERVICE_API_KEY = os.getenv("CONNECTOR_API_KEY")
+app = FastAPI()
+client = CustomClient()
+
+
+def authenticate(Authorization: str = Header(None)) -> None:
+    """
+    Authenticate the user using the 'Authorization' header.
+
+    Args:
+        Authorization (str, optional): Authorization header. Defaults to Header(None).
+    """
+    if Authorization is None or Authorization != f"Bearer {SERVICE_API_KEY}":
+        raise HTTPException(status_code=401, detail="Unauthorized")
+    logging.debug("authenticate: (OK)")
+
+
+@app.post("/search", response_model=Response)
+async def search(request: Optional[Request] = None, 
+                 user: None = Depends(authenticate)) -> Response:
+    """
+    Search Endpoint
+
+    Args:
+        request (Optional[Request], optional): Request object. Defaults to None.
+        user (None, optional): User object. Defaults to Depends(authenticate).
+
+    Returns:
+        Response: Response object (see datamodels.py)
+    """
+    if request is None:
+        logging.debug(f"search: missing_body")
+        return Response(results=[])
+    try:
+        logging.debug(f"search:request.query: {request.query}")
+        data = client.search(query=request.query)
```

Comment:
So currently the client takes care of all the search functionality, some considerations:
- What happens when you need to perform additional calls? For example, sometimes GET calls will need to be made to the service to retrieve some additional required metadata
- Where will serialization live? You've defined a DataItem which is amazing, but should the client take care of that serialization logic? 

Generally our solution was to add a handler file called `provider.py` where all this logic lives, but I'm open to your thoughts

Reply:
@tianjing-li I like the idea of using an intermediate layer for this, will add also a `provider.py`